### ISSUE TITLE
[server] Switch to bi-directional communication

### ISSUE THREAD

**Issue by joshuaauerbachwatson** [joshuaauerbachwatson/anyCards#5]
Although the current app is "mostly client server" it has a push-based aspect because the latest game state (and player lists in the early phases) need to be multicast for efficient operation.   In particular, while using http it is hard to eliminate the polling, which is just wasting cycles and bandwidth.  Also, using explicit http POST operations, with my semi-manual JSON serialization and separate declarations in Swift and Go, is a pain in the neck.   

There are "official" gRPC implementations for Go and for Objective-C on IOS and a slightly less official set of Swift bindings for the latter.   My current reading of gRPC documentation suggests that I can use bi-directional channels and mixture of synchronous RPC, asynchronous RPC, and simple protobuf sends to get exactly the communication semantics I want.  The declarative language for rpcs and messages can become the single point of definition for the protocol, with bindings generated for both languages.

The one thing I'm not sure about is the extent to which App Platform would get in my way.   Right now, the server is doing simple http and is relying on App Platform to manage the external URL with simple binding of paths ... presumably, it is converting https to http in the process.   To use the provided gRPC channels out of the box I believe I would need SSL / TLS end-to-end and and would add a custom bearer token protocol on top of that.   I might have to craft my own gRPC custom channel type to get through the impedence mismatches.   This will be a combination of experimentation and chatting up people at DO to get answers.

**Comment by joshuaauerbachwatson** [comment]
If it turns out that App platform does get in the way, and I don't want to engineer around the problem with software, I could also just run a VM (droplet) with the server on it and use standard gRPC channels protected by SSL/TLS.

**Comment by joshuaauerbachwatson** [comment]
When I originally opened this issue I was convinced that gRPC was the solution but a subsequent discussion with Davi DeBarros at DigitalOcean convinced me that the quickest path to an adequate bi-directional solution was to use websockets.  This would likely cover the most important "pain point" by eliminating the polling.   True that I would not get the benefit of stub generation for both Swift and Go, but the protocol is currently simple enough that I don't really need that.

**Comment by joshuaauerbachwatson** [comment]
A tentative plan.

1.  Use `gorilla` at the server side (a `go` implementation of websockets).
2. Initially, just deploy a second server which is pretty much exactly the `gorilla` sample chat application.
3. Initially, just add a chat channel to the basic `anyCards` app which uses the second server.

At this point we have gotten our feet wet with websockets.   The chat channel is probably useful anyway.   Going forward from this point,

- add a schema to the websocket path so that it can handle game state as well as chat and has the minimal authentication based on the app token 
- retire the old server and standardize on the new one

I have so far gotten the new server deployed on app platform.  Some minor changes to the sample program were needed so that the existing HTML access could continue to work.  These were just changes to the websocket URL (`wss` instead of `ws` and also using the `/websocket` route to distinguish this server from the old one).

I tried and rejected writing a new test client using SwiftUI.  The idea was I could steal some more sample code here and there and get the basic functionality working.  And, getting my feet wet with SwiftUI seemed like a good idea too, so maybe we could convert the entire app to use it.   But, too much learning curve.  I can get the basic websocket thing going more easily by being more incremental

**Comment by joshuaauerbachwatson** [comment]
The chat chennel is now working.

**Comment by joshuaauerbachwatson** [comment]
Before trying to merge the chat and game-state channels and standardize on web sockets, I need to get bearer token authentication working with web sockets.   I have digressed a bit into the auth0 integration question but I see that there is a lot of work, potentially, in that item.   So, I may initially use the old burned-in app token rather than tokens provided by auth0.

**Comment by joshuaauerbachwatson** [comment]
It is now clear that web sockets work fine through App Platform and I _could_ start using it to do the push notifications (and perhaps integrated chat).  However, there seems to be a problem in that web sockets don't integrate bearer token authentication very well.   Perhaps this doesn't matter a lot since I can hand-roll this as part of the data structures that are passed in (as before), and just rely on disconnecting those users who fail authentication.   I think there are two designs to consider and I'm not sure which is better.

1.  Limited use of web sockets.   This would involve having two end points, one which is traditional https and the other which upgrades to web socket.   All the stuff that fits into client/server goes to the traditional endpoint, which uses auth0 infrastructure in the recommended way.   Requests that open push or chat channels go to the web socket endpoint but perhaps only after the user has been validated through the traditional endpoint and has been given a one-time secret.  This must be passed when opening the web socket.
2. Figure out how to make the web socket access fully secure and integrated into auth0, after which everything can be built on it.

Note that some decision about how to integrate a bearer token into web socket is common to both.

**Comment by joshuaauerbachwatson** [comment]
> Note that some decision about how to integrate a bearer token into web socket is common to both.

On the Swift side, the `URLSessionConfiguration` can specify headers.   The trick is that they don't want you to specify the `authorization` header.   It seems, however, that people do specify it successfully.   To avoid any later problems, _if_ you control both client and server (as I do), then you can get around any long-term risk by just using an alternate (non-standard) header instead of `authorization`.   That sounds like a plan.

On the `go` side, since `upgrade` takes an `http` request as its argument, you can parse out the header from it and subject that to whatever validation is required.

Despite this being relatively straightforward, I am inclined to use `websocket` only for chat and for push and to use internally generated tokens for validation.   In the short run, the rest of the protocol can use the app token as it does at present.

It would be good to keep the planned `auth0` integration orthogonal and only switch over to it when we are ready to ditch the app token in favor of something generated by `auth0`.

**Comment by joshuaauerbachwatson** [comment]
> I am inclined to use `websocket` only for chat and for push and to use internally generated tokens for validation. In the short run, the rest of the protocol can use the app token as it does at present.

Observations.

1.  There is a difference between using two endpoints and actually having two servers.   Right now I have two servers for expedience in quickly exploiting the `Gorilla` sample code, but I was getting charged twice as much monthly by Digital Ocean as a consequence, so I shut down the second server to save money while I was not really working on AnyCards.   My plan now is to use the `Gorilla` sample only as education and a possible cut-and-paste source but to integrate all new functionality into the existing server.   The `go` SDKs should easily handle this.
2.  Once a player has joined a game, he _must_ have the `websocket` channel for game state pushes and should be free to start using it for chat.   So, the entire first-connection sequence should be integrated into one module.   With this approach it is easy to generate a one-time random ID token for each user.   

> It would be good to keep the planned `auth0` integration orthogonal and only switch over to it when we are ready to ditch the app token in favor of something generated by `auth0`.

Since I am already part way down the path with `auth0` and have learned a little more about how to complete the authentication handshakes on the server side, I think it will be more efficient to work on the two changes together.  I have therefore closed issue #13.

**Comment by joshuaauerbachwatson** [comment]
The schema for what is sent on the bi-directional channel can be nearly trivial.  

- Every transmission is just a type token followed by either a chat contribution or a game state.   
- In the upload direction, the only valid type token is 'chat' (otherwise, we would need to complicate the protocol by adding responses; the idea is to use the old http based channel for all request/response flows).
- A chat contribution is just a string.  It takes the form`<User>: <Message>` where the user is the id of the originator.  Chat contributions can flow either way.
- A game state only flows in the download direction and it consists of an encoded game state,  I won't worry now about whether the game state schema is promoted to be part of this schema or whether it starts At pas on opaque string and goes through a second decode.

I would expect the schema on the request/response channel to remain largely the same, in the short run, other than 
- removing the poll/response sequence
- retiring the app token from the individual messages, in favor of using `auth0` for all security

At present, I plan to retain "Nearby Only" mode because, if anything, the two communication styles are now better aligned semantically than they were before.   If I need to ditch it, I will.  If I keep it, I should consider some logic to sort out multiple ongoing games in the same "proximity."   RIght now, the protocol assumes that everyone who is running AnyCards in proximity is in the same game.

**Comment by joshuaauerbachwatson** [comment]
Implementation sequence.

- [ ] Complete the `auth0` integration on the existing https channel and drop the app token.
- [ ] Create the second endpoint and figure out how it gets authenticated and integrated into the `auth0` handshake.  Don't have it do anything yet.
- [ ] Institute the new protocol on the bi-directional channel and start using it for chat.
- [ ] Switch over the "new game state" logic to use the new channel.  Try to align the two communication types better.
- [ ] FIgure out what to do about chat in Nearby mode.   Perhaps it is disabled?

**Comment by joshuaauerbachwatson** [comment]
Item one in the task list is _almost_ done and what's been coded is actually working.   We have indeed dropped the appToken and auth0 is now verifying that its tokens are legal.  I need to add some verbosity to the server so I am sure the auth0 code is doing the right thing.  Then, I need to add a scope for the admin operations (reset and dump) so that normal user logins do not get conferred those privileges.

Currently, auth0 access tokens are set to expire in 1 week (seems reasonable).  There were some issues with the `expiresIn` field because the Auth0 Swift library seems to do "smart" decoding (adds the interval to "now") but dumb encoding (just writes out the date).  The "smart" decoding is not actually smart enough to know that it is getting a date rather than an interval, so the `expiresIn` value becomes incorrect after one round-trip to disk and back.   I believe I fixed that.

**Comment by joshuaauerbachwatson** [comment]
I found that I could not easily use scope checking as recommended in the auth0 quickstart.   I really wanted to confer the permission using the auth0 dashboard based on user identity (basically, just give it to myself).   Doing that in the naive way, using RBAC, I could not figure out how to convert permissions into scopes, but it was possible to have the permissions appear in the token.  So, I simply changed the auth0 sample code so that it defined the CustomClaim as consisting of the permissions rather than the scopes.

Anyway, I now regard the first item in the task list as complete.

**Comment by joshuaauerbachwatson** [comment]
For the second item, it's clear we need a websocket for each player and the players of a single game have to have their websockets associated with the game (chats are scoped to games and push notifications are organized around games).   To facilitate the setup, we need a `join` endpoint which presents a game token and gets back a unique secret for use in initiating the websocket.

**Comment by joshuaauerbachwatson** [comment]
I have tentatively completed the integration of the Gorilla sample into the server but the logic is currently incomplete.   There is only one `Hub` object, and it is created during server initialization.  This means that everyone connected to the server will be in the same chat instead of a chat per game.   I need to eliminate the global `Hub` variable and add a `Hub` member to the stored game state.   The creation of hubs, the creation of clients for new players, and the joining of clients to hubs all needs to occur later as players join games and games are created to receive them.

**Comment by joshuaauerbachwatson** [comment]
Sketch of something that might work:

- When a `Game` is created in the server, its `Hub` is created also.
- At the client end (currently, just when opening a chat), a `URLRequest` is created with added headers for player (order string) and game (token).  Also add the standard authorization header with the proper bearer token for auth0.
- In the handler for the `/websocket` route first validate as with any auth0 request, then retrieve the headers for game and player to find the correct hub.   Create a client for that player and add it to the hub.

One subtlety is knowing whether the player already has a client.   We could use a variety of indicators but we need to know when a client becomes invalid (unregistered) because its connection failed.

**Comment by joshuaauerbachwatson** [comment]
> One subtlety is knowing whether the player already has a client. We could use a variety of indicators but we need to know when a client becomes invalid (unregistered) because its connection failed.

Sketch of something that might work
- Consolidate all the termination logic in `Client` so that all failures close the connection, terminate the ticker. cause all goroutines to exit (and, of course, unregister).
- Add a `()->Void` handler to the `Client` which is always called when the Client terminates.
- Use a simple boolean in the `Player` structure to say whether there is currently a Client.  Set this true when a new client is created and false in the handler callback.

**Comment by joshuaauerbachwatson** [comment]
I have something coded and I pushed it because I need it running in DO in order to test.   But, I'm only now going to find out if it works.

**Comment by joshuaauerbachwatson** [comment]
It doesn't work so far.    The authorization header does seem to be processed. correctly so that part of the design is good.  The custom headers being used to convey the playerOrder and the gameToken are not arriving at the server (or so it seems).  I could try to get to the bottom of that or just switch to passing those in the query.   Yes, the query is less secure but these are not secrets.

**Comment by joshuaauerbachwatson** [comment]
Ok.  Switched to using the query instead of headers and chat now starts up correctly and my messages get properly echoed.   Yes, I will need to test with multiple players but I think first I should work on eliminating poll and using the bi-directional channel for push notifications.

**Comment by joshuaauerbachwatson** [comment]
In eliminating `poll` I need to remember that the current `poll` serves three purposes.

1.  It keeps every app up to date with the current `GameState`.   That is the one thing that will be subsumed by using the bi-directional channel for push notifications.
2. It is a liveness test for the app.   This will need to be done in some other way.
3. The first poll causes the `Game` and `Player` structures to be initialized.  It is only then that the websocket can be opened.   This will not be subsumed and should be retained, with `Poll` being renamed to `Join`.

**Comment by joshuaauerbachwatson** [comment]
> 2\. It is a liveness test for the app.   This will need to be done in some other way.

The gorilla code that I added to the server is doing the recommended ping/pong protocol as defined by the RFC.  What is not completely clear is whether my client code (using `URLSessionWebSocketTask` is automatically responding to pings.  There is certainly no explicit code there and no obvious way I can see to add a handler for incoming control frames such as ping frames.   There is a way to _send_ pings and specify a pong handler but that is not what I need or want.

Assuming there is an automatic response, then what I would need at the server end is either

1.  Revise the age-out mechanism at the server end to stop relying on per-player idle counts.   As it stands, only a transmitted game state will zero out this value.  A user who is just thinking or watching other players play can time out doing so.
2. Verify that the pong handler at the server gets called on productive pings as well as unproductive ones and use that handler to zero out the appropriate idle count.

If there is not an automatic response then I need further reaching changes.  Either

1.  Get a better client library, one that does automatic pings.
2. Stop sending pings from the server and instead rely on ones from the client.

**Comment by joshuaauerbachwatson** [comment]
I am sure (based on documentation) that the server end could zero out the idle count in a pong handler and so I'm doing that.
I an not sure (will need to determine empirically) whether the client end will automatically send the pong.  So, I would need to do more work if it doesn't.  If it does, I've got the liveness issue covered.

**Comment by joshuaauerbachwatson** [comment]
While I've got this whole area open, it occurs to me that I could simplify things, perhaps radically.

- There might not be need for a `join` since the act of opening the websocket channel provides enough opportunity to initialize the game and player structures.
- That leaves `newState`, `withdraw`, `dump`, and `reset`.    The last two are administrative and the app doesn't issue them, so they are somewhat outside the protocol.   That leaves `newState` and `withdraw`.
- A player can `withdraw` just be closing the websocket channel, which is then detected at the server end.   In fact we want to make sure that state is being cleaned up when any kind of failure happens on this channel.   That leaves `newState`.
- Just as we distinguish chat messages and game states in the server->client direction, we could also distinguish them in the client->server direction.   That allows the app to use _just_ the websocket channel and no longer any https requests.

There's a pleasing simplicity and symmetry to this approach.   I would keep the `dump` and `reset` endpoints for administration purposes.   I currently just use scripted `curl` to drive these anyway.

**Comment by joshuaauerbachwatson** [comment]
Although it could leave some mess, I would try to implement the previous comment's sketch incrementally, eliminating `poll`, `withdraw`, and `newState` in that order.

**Comment by joshuaauerbachwatson** [comment]
I believe I have the client-side communicator revised to use the web socket for game state and eliminating poll, but the chat controller still thinks it has its own web socket and the server side still needs revision.   This will be a fairly large change by the time the system is stable enough for the next commit.

For chat it makes sense for the transcript of incoming messages to be maintained in `ViewController` and for the chat window to be primed with it every time it is opened.   The `ViewController.newChatMsg` function needs to be sensitive to whether the chat window is being displayed and keep it in synch.

We also need to be sure that the `Chat` button is not displayed when using nearby mode.

**Comment by joshuaauerbachwatson** [comment]
Client side of poll elimination is tentatively done and committed on branch `useWebsocket`.  Can't really test until the server side is updated.

**Comment by joshuaauerbachwatson** [comment]
Tentatively finished server side; in order to test I went for broke and both merged and pushed work to date.

Things to test before moving on
* [ ] Game state propagation is working (nominal game play with two players)
* [ ] Nearby mode still works (not broken)
* [ ] Chat works, including the new functionality whereby the transcript is kept permanently even when chat window is not showing.

**Comment by joshuaauerbachwatson** [comment]
So far game state propagation is not working either for server based or for nearby.  And, the chat button hide flag is inverted (it does not show in server based where it should, and it does show in nearby, where it shouldn't).  Chat issue is probably trivial.   Then, maybe I should work on nearby because that is pure regression (no new functionality there).   Then, I will needs lots of instrumentation to go after why server-based doesn't work.

**Comment by joshuaauerbachwatson** [comment]
The breakage in nearby mode was due to a flaw in `ViewController.doGameChanged` that probably affects server based also.  Nearby mode is now working.

**Comment by joshuaauerbachwatson** [comment]
Server-based game state propagation is broken more profoundly than nearby (the `doGameChanged` function is never called).   I think I overlooked yet another function that `poll` was performing, which was to drive the initial sharing of player information.   No one is sending player lists and so no progress can be made.   At least, that's a plausible hypothesis.

**Comment by joshuaauerbachwatson** [comment]
The following fix might work
- After creating the server-based communicator, immediately send the current state on it.
     - this requires some thought and possible redesign, since the multipeer communicator isn't designed to work that way
- Upon getting a new state that doesn't include a complete stable player list, send current state again
     - this is a subtle variation on what is there now, which sends current state only when the player list changes; what's there now is mostly in service of multipeer
     - there is a possible problem of too rapid an exchange of messages prior to convergence.  What we had before, based on `poll`, had a two-second wait between retries.

Alternatively, we could resurrect the poll message, call it `join` (after all) and run the two-second timer as before.  The timer would be stopped once the player list converged.

**Comment by joshuaauerbachwatson** [comment]
As I discussed in issue #31, I think we should more cleanly separate player list management from game state propagation and abstract most of the player list management in the communicator since it really differs a lot when using peer versus server.   The game is ready to start (and game states start to propagate) only when the player list is complete.

Part of this is removing the player list from the game state.   Game state propagation should be just that.   Both communicators should use separate callbacks (consistently) for changes in game state and changes in the player list.   The messages to manage the player list should not be exposed outside the communicators.

There are three tasks, here, although it might be hard to sequence them strictly.

* [ ] Redesign `Communicator` and `CommunicatorDelegate` with suitable abstraction and make sure the logic above the `Communicator` is clear and consistent.
* [ ] Design a new protocol for Nearby to use
* [ ] Design a new protocol for Entire Internet to use

It makes sense for the server to be the canonical source of player lists for Entire Internet and the leader to have that role in Nearby.

**Comment by joshuaauerbachwatson** [comment]
### Redesign `Communicator` and `CommunicatorDelegate`

- Remove `updatePlayers` from `Communicator`.   Player management moves are never initiated from outside the Communicator.   Construction of the `Communicator` is sufficient to start the player management protocol.
- Replace `connectedDevicesChanged` in `CommunicatorDelegate` with `newPlayerList` or something of that nature.   This call should include the number of players (meaningful for everyone except the leader) plus an array of Players.
- I think we keep everything else.

**Comment by joshuaauerbachwatson** [comment]
### Redesign `ServerBasedCommunicator`

- In the server->client direction, add a type 'P' message frame with a player list (the type 'G' frame will no longer include players).   Also add a type 'L' frame for lost player.
- Let the server notice whenever a new player registers for a game and send a type 'P' frame to everyone.  Similarly, let the server notice whenever a player leaves a game (or fails out of a game) and send a type 'L' frame to everyone.
- Management of game state is as before.

**Comment by joshuaauerbachwatson** [comment]
I should probably first implement (in the client) the interface changes and the changes above them plus the changes to the server communicator, keeping the multipeer communicator compiling but not expecting it to work.   Then, implement the server changes to support those client changes.   During this process there will likely be design changes.   At that point, I'll be ready to design the new multipeer.  I think my memory of exactly how multipeer works is not good enough right now.

**Comment by joshuaauerbachwatson** [comment]
As I start to work on this, I see that there are certain errors (too many players, too many leaders) that are being caught in the client.  Since the client will respond to any such error by disconnecting, I think we can leave this in place rather than putting the responsibility below the communicator interface.

**Comment by joshuaauerbachwatson** [comment]
The next difficulty comes from the fact that the server does not now store player names, only order numbers.   While the state structure is easily amended to fix this, transmission of the values becomes more complicated.   The user name has no syntax restrictions preventing blanks or colons (separators I would like to use).   For the moment, I will base64-encode the user name (this is likely to work fine but I will revisit if needed).

**Comment by joshuaauerbachwatson** [comment]
On local branch `newCommunicatorProtocol` I've saved a revised `Communicator` and `CommunicatorDelegate` definition along with a tentative rewrite of the logic above the communicator line.  While almost all the responsibility for player list management is now in the communicators, the main controller still detects the conditions of "too many players", "too many leaders", and "no leaders", as well as reacting when there exactly enough players.   I changed the actual communicators only enough to keep them building but I started down the line of designing the player list message that will be transmitted from server to client.

**Comment by joshuaauerbachwatson** [comment]
Now that the server is aware of the number of players and this is being sent in messages, the semantics of this value is a little muddy.   Requirements:

- only the leader shall specify the number of players.  The leader knows the number by the time a communicator is started.
- the server either knows the number of players or not.  Once the number becomes known it doesn't change
- all clients other than the leader are in the same position as the server and only know the number of players if they have received it

So, if you are not the leader, don't try to tell the server the number of players.
If you are not the leader, don't read the number of players from `OptionSettings` (might be a stale value from the last time you were leader).  Instead, always return 0 unless the value has been set by a callback from the communicator.

**Comment by joshuaauerbachwatson** [comment]
Starting to test with new code.   Two problems so far

1.  For the second time I've observed a hard failure of auth0 login with _Login failed with An unexpected error occurred. CAUSE: Error Domain=com.apple.AuthenticationServices.WebAuthenticationSession Code=3 "Scheme '(null)' is not valid as per RFC 3986." UserInfo={NSDebugDescription=Scheme '(null)' is not valid as per RFC 3986.}_.   When I simply retried, the login succeeded the second time and we moved on to the next error.   I've only seen this on the mac (never on an iOS device).
2. There was a "malformed request" response from the server when creating the websocket.   This gets treated as a fatal error which makes it a pain to track down.   It should be logged as a normal exception and the user given some sort of pop-up.
3. The malformed request occurred because the player token was not tokenized, and yet it got stringified and sent anyway.  Using string interpolation and `player` instead of `player.token`, what was sent was the debug string for the player.

**Comment by joshuaauerbachwatson** [comment]
Added more logging, fixed some more minor protocol errors.   The next major bug found was that I was running the Hub synchronously (blocking on it) rather than using a goroutine.

**Comment by joshuaauerbachwatson** [comment]
Next:

1.  Even though my client is leader it does not appear to be sending the number of players.
2. When a game or client is cleaned up, a lost player message may be sent on a web socket which is then immediately closed; at the client end there is a mess.  Termination sequences need to be more carefully worked out.
3. Server does not start the player list with numPlayers (this is due to misunderstanding how to turn integers into strings in golang).
4. Logging of incoming frames on the websocket using integer codes rather than recognizable names.

**Comment by joshuaauerbachwatson** [comment]
After quite a bit of cleanup, it looks like a single client connecting to the server works in accordance with the next protocol, except that the server cleanup thread ends up reaping the client connection.   This means something is going wrong with the server-initiated ping/pong protocol, which should be reseting the idle count for the player.

**Comment by joshuaauerbachwatson** [comment]
> This means something is going wrong with the server-initiated ping/pong protocol, which should be reseting the idle count for the player.

The problem may not be in the logic but rather in the parameterization.  The player cleanup period is only 45 seconds but the ping interval is 54 seconds.   Clearly, the player cleanup period needs to be at least 2-3 times the ping interval.

There is a logic problem also in that the pong handler is only resetting the player idle count and not the game idle count.   The direct fix for this is to give either the player or the websocket client a reference to the game.   Before embarking on that change, it is worth considering whether a game timeout is even needed when there are player timeouts.

One alternative is to let the cleanup thread notice whether a game is complete (has all its players).  If not, it increments the idle count.  If so, it zeros the idle count and leaves the game alone.  A consequence is that the game timeout is now governing how long to wait for the players to assemble.    Once they've assembled, the game stays.

Games are already being deleted when the last remaining player leaves.   I am not sure I have a clear position on whether a player who leaves a game can rejoin it.  If so, then the game idle count should resume once a single player has left.  If not, the game should be deleted (and all client connections closed) once a single player leaves.   I think what we are doing now comes down to nearly the same thing, because each app will react to a lost player message by closing its connection to the game.

**Comment by joshuaauerbachwatson** [comment]
Now I seem to get through the player list coordination and two instances believe they are playing with each other.

The lead player is sending game state, the server is is forwarding it, and log messages suggest that the other player is receiving it.  But, the non-lead player's screen does not update.

Also, the server cleanup is clearly not quite right.

**Comment by joshuaauerbachwatson** [comment]
The app was still expecting to get `ReceivedState` (now obsolete) instead of `GameState`.   Fixing that makes the communication basically work.

So, I'm declaring victory on the poll elimination steps.   Still remaining:

- do whatever redesign is needed for nearby mode so I can check off that item.
- fix cleanup at server end and termination at client end (when you press End Game it works but you get a gratuitous communication failure popup).
- eliminate newstate and withdraw endpoints (which means that we start sending things other than chat on the websocket upbound).

**Comment by joshuaauerbachwatson** [comment]
Nearby mode is working again.   I'm not posting a full-fledged design.  Since we use Player tokens as peer ids it is relatively easy to get the player list management to work.   But, I did have to arrange to propagate the number of players from the leader to the other players.   I do this using the `discoveryInfo` field (this is a key-value list and can mean whatever you need it to mean ... in this case, I have a single key for transmitting the number of players.  Only the leader sends this).

**Comment by joshuaauerbachwatson** [comment]
> ... let the cleanup thread notice whether a game is complete (has all its players). If not, it increments the idle count. If so, it zeros the idle count and leaves the game alone.

This seems like the right direction so I'll do that.

> Games are already being deleted when the last remaining player leaves. 

They should be.  I need to check whether that's true.  I think it's adequate because as soon as the server sends out a "lost player" message all the remaining players will withdraw.   So, sooner or later the game reaches zero players and gets deleted.

**Comment by joshuaauerbachwatson** [comment]
I believe cleanup is now switched to the new scheme but I have to test the various scenarios.   Meanwhile, I realize I never tested whether chat works, so I will do that today.

**Comment by joshuaauerbachwatson** [comment]
> Meanwhile, I realize I never tested whether chat works, so I will do that today.

Yup.  Chat is broken.  The server is assuming that chat messages arrive with a message type, so it immediately re-broadcasts without formatting a message.    Meanwhile, the client is still assuming that only chat messages are sent, so it doesn't add a message type. 

I could resolve this either way but since we are about to start sending more message types from the client I should fix this at the client end.   The server end should be fixed to expect a type that is valid for an incoming message (either chat or game state).   At the same time, I might as well change the client to actually send game state on the websocket.

This gets us pretty close to the enumerated goals for this issue.   The final step will be to eliminate the withdraw message and then clean up both client and server to expect that all "normal" communication will be via the websocket (the server will still validate admin priviledge and carry out admin requests via https).

**Comment by joshuaauerbachwatson** [comment]
I tested cleanup.  The server-end logic is sound as far as it goes but it doesn't take into account that clients will try to send withdraw messages and this results in the game being re-established momentarily while each client withdraws!   This is messy and probably involves some race conditions that will bite me in the ass eventually.   So, I will eliminate the withdraw message sooner rather than later.  This suggests going non-incremental for the rest of the work.   Part of eliminating withdraw is being a little more conscious of how the websocket is terminated:  at the server end, using Gorilla, I need to send a close message explicitly.

Something else that needs doing: be very explicit about the lifecycle of a game.  I believe a game is shorter-lived now than it once was.  It starts out being incomplete (not enough players or unknown number of players).  From there, it either becomes complete or is cleaned up.  Once the game is complete, it lasts until some connection breaks (either because of an error or because the client withdraws), at which point is should be cleaned up aggressively.

**Comment by joshuaauerbachwatson** [comment]
As I revise the protocol I see that it is no longer necessary to store the game state on the server.  Game states are just multicast.  This multicasting cannot begin until all players are assembled so there is no longer a need to "catch up" when joining the game late.

**Comment by joshuaauerbachwatson** [comment]
I think the main things that remain are
- debugging
- making sure the termination sequences are clear and race-free

**Comment by joshuaauerbachwatson** [comment]
It's clear that termination is still messy but it's also clear that getting the termination flow exactly right is not so easy either, and the difficulties vary by which SDK you are using.  I am using different ones in different languages at each end.  So, the plan is to just do general debugging and focus on making it clean from a user perspective (no annoying gratuitous popups in the client, no distracting error reports in the server transcript).   If we don't follow the RFC exactly, no biggie, as long as things keep working.

**Comment by joshuaauerbachwatson** [comment]
The first small fix is to detach a terminating client from the hub _before_ sending the lost player message.  This ensures that there is no attempt to send the message to the terminating client itself.

**Comment by joshuaauerbachwatson** [comment]
By virtue of filtering certain "disconnected" type errors, termination is now superficially clean.  Whether it is always canonical underneath is not certain but I think it does not matter (until it does).

There were other bugs that were only reveals when actual play starts.  The most noteworthy is that it requires a very large maximum message size to handle game states when they are serialized using JSON with default coding keys.   Since we support decks of up to 106 cards (double deck with two jokers), and it seems like encoding comes out to about 80 bytes per card, we need a 10K limit (not 0.5K as we started out with).

We could shrink this somewhat without abandoning the convenient JSON encoding by using artificially short coding keys.

The next move beyond that would be to move to a binary format.  The difficulty there is that we would need both a Swift and a Go implementation.   This is probably available with protobuf but that's a big step.   I think for safety and ease of development we would need a tagged format (the data types are just complex enough, though not actually recursive, that I would not try for an untagged format).

**Comment by joshuaauerbachwatson** [comment]
I decided it was time to declare victory on this long-lived work item.  Some things may have regressed, which I need to keep an eye out for.  Some concerns have been identified (large message sizes when transmitting game states) which I should probably open an issue for.   But, I need to move forward into actual play sequences with multiple players and come back to protocol issues when necessary.
